src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,325 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "feat/wave-reader.h"
+
+namespace kaldi {
+
+void ElementwiseProductOfFft(const Vector<BaseFloat> &a, Vector<BaseFloat> *b) {
+  for (int i = 0; i < a.Dim() / 2; i++) {
+    float tmp = (*b)(2*i);
+    (*b)(2*i) = tmp * a(2*i) - (*b)(2*i+1) * a(2*i+1);
+    (*b)(2*i+1) = tmp * a(2*i+1) + (*b)(2*i+1) * a(2*i);
+  }
+}
+
+void ConvolveSignals(const Vector<BaseFloat> &filter, Vector<BaseFloat> *signal) {
+  int signal_length = signal->Dim();
+  int filter_length = filter.Dim();
+
+  int fft_length = 2;
+  while (fft_length < signal_length + filter_length - 1) {
+    fft_length = fft_length * 2;
+  }
+  KALDI_VLOG(1) << "fft_length for full signal convolution is " << fft_length;
+
+  SplitRadixRealFft<BaseFloat> *srfft_;
+  srfft_ = new SplitRadixRealFft<BaseFloat>(fft_length);
+
+  Vector<float> filter_padded(fft_length);
+  filter_padded.Range(0, filter_length).CopyFromVec(filter);
+  srfft_->Compute(filter_padded.Data(), true);
+  Vector<float> signal_padded(fft_length);
+  signal_padded.Range(0, signal_length).CopyFromVec(*signal);
+  srfft_->Compute(signal_padded.Data(), true);
+
+  ElementwiseProductOfFft(filter_padded, &signal_padded);
+
+  srfft_->Compute(signal_padded.Data(), false);
+  signal_padded.Scale(1.0 / fft_length);
+
+  signal->CopyFromVec(signal_padded.Range(0, signal_length));
+}
+
+void BlockConvolveSignals(const Vector<BaseFloat> &filter, Vector<BaseFloat> *signal) {
+  int signal_length = signal->Dim();
+  int filter_length = filter.Dim();
+
+  KALDI_VLOG(1) << "Length of the filter is " << filter_length;
+
+  int fft_length = 2;
+  while (fft_length < 4 * filter_length) {
+    fft_length = fft_length * 2;
+  }
+  KALDI_VLOG(1) << "Best FFT length is " << fft_length;
+
+  int block_length = fft_length - filter_length + 1;
+  KALDI_VLOG(1) << "Block size is " << block_length;
+  SplitRadixRealFft<BaseFloat> *srfft_;
+  srfft_ = new SplitRadixRealFft<BaseFloat>(fft_length);
+
+  Vector<float> filter_padded(fft_length);
+  filter_padded.Range(0, filter_length).CopyFromVec(filter);
+  srfft_->Compute(filter_padded.Data(), true);
+
+  Vector<float> temp_pad(filter_length - 1);
+  temp_pad.SetZero();
+  Vector<float> signal_block_padded(fft_length);
+
+  for (int po = 0; po < signal_length; po += block_length) {
+    // get a block of the signal
+    int process_length = block_length;
+    if (signal_length - po < process_length) process_length = signal_length - po;
```

Comment:
To me, it would be a bit easier to read as 
```
if (process_length > signal_length - po) process_length = signal_length - po;
```
or, which arguably reads better,
```
int process_length = std::min(block_length, signal_length - po);
```